File: examples/matlab/mfiles/f_VertexFunctionOperations.m
%% Operations on vertex functions

clear ; close all ; path=pwd();
[s,e]=regexp(path,'.+(?=[\\/]{1,2}examples)');
path=path(s:e); addpath(genpath(path)); clear s e path
%% 1) Scalar case
% Elementary |VertexFunction| can be combined with others, the computation of 
% the expression and derivative is done automatically.
% a) Some operations

f1s = VertexFunction(@(a) a.^2, @(a) 2*a, "f1s");
f2s = VertexFunction(@(a) a.^3, @(a) 3*a.^2, "f2s");
f3s = VertexFunction(@(a) 0.1*a, @(a) 0.1*ones(size(a)), "f3s");
%% 
% * Operations with scalars

f1s = f1s*9+pi
% 6+9*f1s does not work, the numbers should be placed after the
% VertexFunction
%% 
% * Operations with other |VertexFunction|

f2s = f1s.*(f2s/3+f3s*2).^(-f2s)
f3s = f1s./f3s

f1s.plot()
% b) Interpolation
% a) Definition of the interpolation domain

defInterpScalar.Domain = Domain(3);
defInterpScalar.Children = [f1s,f2s,f3s];
defInterpScalar.Label = "interpScalar";
P = Penalization("simp",2);
defInterpScalar.Penalization = P;
P.plot();

interpScalar=Interpolation(defInterpScalar);
% b) Coordinates of in the polygon

n = 1e3; % number of evaluation points
x = interpScalar.initializeVariable(n,"rand",0.2)
x = interpScalar.projection(x);
figure()
interpScalar.plot(x);
% c) Evaluation
% Each point $x$ is associated to a scalar field value $a$.

a = rand(1,1,n);
[w,dwdx] = interpScalar.evalBasisFunction(x);

tic
val = interpScalar.eval(x,a,w); % value
tv1 = toc;
disp(strcat("Compute interpolated values in ",num2str(tv1*1000)," ms"))

tic
dvalda =  interpScalar.evalda(x,a,w); % derivative w.r.t a
tda1 = toc;
disp(strcat("Compute interpolated a-derivative in ",num2str(tda1*1000)," ms"))

tic
dvaldx =  interpScalar.evaldx(x,a,w,dwdx); % derivative w.r.t x
tdx1 = toc;
disp(strcat("Compute interpolated x-derivative in ",num2str(tdx1*1000)," ms"))
% Check the Taylor expansions

epsilon = logspace(-8,-2,10); % test for different epsilon
res = zeros(10,1,n);

for i=1:10
    pert = epsilon(i)*rand(size(a));
    valPerturbeda = interpScalar.eval(x,a+pert,w);
    res(i,1,:) = vecnorm(valPerturbeda - (val + mult(dvalda,pert)),2,1);
end
MaxRes = max(res,[],3); medRes = median(res,3);
loglog(epsilon,medRes,'-o',epsilon,MaxRes,'-o',epsilon,epsilon.^2,'k--');
legend("med res","max res","\epsilon^2","Location","northwest")
xlabel("\epsilon")
ylabel("Norm of Taylor remainder")
title("Taylor remainder value w.r.t a")
grid on

res = zeros(10,1,n);
l = defInterpScalar.Label;

for i=1:10
    pert = epsilon(i)*rand(size(x.(l)));
    xPert.(l) = x.(l) + pert;
    valPerturbedx = interpScalar.eval(xPert,a);
    pert = reshape(pert.',2,1,[]);
    res(i,1,:) = vecnorm(valPerturbedx - (val + mult(dvaldx.(l),pert)),2,1);
end
MaxRes = max(res,[],3); medRes = median(res,3);
loglog(epsilon,medRes,'-o',epsilon,MaxRes,'-o',epsilon,epsilon.^2,'k--');
legend("med res","max res","\epsilon^2","Location","northwest")
xlabel("\epsilon")
ylabel("Norm of Taylor remainder")
title("Taylor remainder value w.r.t x")
grid on
%% 2) Vector case
% Only inner product is implemented.

dimInput = 2;  dimOutput = 2;
f1v = VertexFunction(@(b) b.^2, @(b) [2*b(1,1,:),zeros(1,1,size(b,3));zeros(1,1,size(b,3)),2*b(2,1,:)], "b.^2", dimInput, dimOutput);
f2v = VertexFunction(@(b) b.^3, @(b)  [3*b(1,1,:).^2,zeros(1,1,size(b,3));zeros(1,1,size(b,3)),3*b(2,1,:).^2], "b.^3", dimInput, dimOutput);

f3v = f1v.innerProduct(f2v) .* f2v
f1v = f3v *2;
f2v = f3v +4;
f2v.plot()
% b) Interpolation
% a) Definition of the interpolation domain

defInterpVec.Domain = Domain(3);
defInterpVec.Children = [f1v,f2v,f3v];
defInterpVec.Label = "interpVec";
P = Penalization("simp",2);
defInterpVec.Penalization = P;
P.plot();

interpVec=Interpolation(defInterpVec);
% b) Coordinates of in the polygon

n = 1e3; % number of evaluation points
x = interpVec.initializeVariable(n,"rand",0.2)
x = interpVec.projection(x);
figure()
interpVec.plot(x);
% c) Evaluation
% Each point $x$ is associated to a vector field value $b$.

b = rand(2,1,n);
[w,dwdx] = interpVec.evalBasisFunction(x);

tic
val = interpVec.eval(x,b,w); % value
tv1 = toc;
disp(strcat("Compute interpolated values in ",num2str(tv1*1000)," ms"))

tic
dvaldb =  interpVec.evalda(x,b,w); % derivative w.r.t a
tda1 = toc;
disp(strcat("Compute interpolated a-derivative in ",num2str(tda1*1000)," ms"))

tic
dvaldx =  interpVec.evaldx(x,b,w,dwdx); % derivative w.r.t x
tdx1 = toc;
disp(strcat("Compute interpolated x-derivative in ",num2str(tdx1*1000)," ms"))
% Check the Taylor expansion

epsilon = logspace(-8,-2,10); % test for different epsilon
res = zeros(10,1,n);
for i=1:10
    bpert = epsilon(i)*rand(size(b));
    valPerturbedb = interpVec.eval(x,b+bpert);
    res(i,1,:) = vecnorm(valPerturbedb - (val + mult(dvaldb,bpert)),2,1);
end
MaxRes = max(res,[],3); medRes = median(res,3);
loglog(epsilon,medRes,'-o',epsilon,MaxRes,'-o',epsilon,epsilon.^2,'k--');
legend("med res","max res","\epsilon^2","Location","northwest")
xlabel("\epsilon")
ylabel("Norm of Taylor remainder")
title("Taylor remainder value w.r.t b")
grid on


res = zeros(10,1,n);
l = defInterpVec.Label;

for i=1:10
    pert = epsilon(i)*rand(size(x.(l)));
    xPert.(l) = x.(l) + pert;
    valPerturbedx = interpVec.eval(xPert,b);
    pert = reshape(pert.',2,1,[]);
    res(i,1,:) = vecnorm(valPerturbedx - (val + mult(dvaldx.(l),pert)),2,1);
end
MaxRes = max(res,[],3); medRes = median(res,3);
loglog(epsilon,medRes,'-o',epsilon,MaxRes,'-o',epsilon,epsilon.^2,'k--');
legend("med res","max res","\epsilon^2","Location","northwest")
xlabel("\epsilon")
ylabel("Norm of Taylor remainder")
title("Taylor remainder value w.r.t x")
grid on

File: examples/python/py/f_VertexFunctionOperations.py
# -*- coding: utf-8 -*-
"""
Operations on VertexFunctions
"""

import matplotlib.pyplot as plt
import numpy as np
import sys, os
sys.path.append(os.path.relpath("..//..//..//src//python"))
import hiped as hp

#%% 1) Scalar case
'''
Elementary VertexFunctions can be combined with others, the computation of the 
expression and derivative is done automatically.
'''
#%% a) Some operations

f1s = hp.VertexFunction("f1s", lambda u: u**2, lambda u: 2*u)
f2s = hp.VertexFunction("f2s", lambda u: u**3, lambda u: 3*u**2)
f3s = hp.VertexFunction("f3s", lambda u: 0.1*u, lambda u: 0.1*np.ones(u.shape))

# Operations with scalars

f4s = (2*f1s + 3)**2
print(f4s)

# Operations with other VertexFunctions

f5s = f1s*(f2s/3+f3s*2)**(-f2s)
print(f5s)
f6s = f1s/f3s
print(f6s)

#%% b) Interpolation

n = 100
penal = hp.Penalization("ramp", 3)
interpScalar= hp.Interpolation(domain = hp.Domain(6), children = [f1s,f2s,f3s,f4s,f5s,f6s],
                               label = "interpScalar", penalization = penal)


x = interpScalar.setInitialVariable(n, "rand", radius = 1) # initialization of the variables
x = interpScalar.projection(x) # projection onto the domain
u = np.random.rand(1,1,n)
plt.figure()
interpScalar.plot(x)
plt.show()

#%% c) Evaluation

w, dwdx = interpScalar.evalBasisFunction(x)
f = interpScalar.eval(x, u, w) # value
dfdu =  interpScalar.evaldu(x, u, w) # derivative w.r.t a
dfdx =  interpScalar.evaldx(x, u, w, dwdx) # derivative w.r.t x

#%% Check Taylor expansion - u

from hiped.utils import mult

# derivative with respect to u
h = np.logspace(-8,-2,10) # test for 10 different h
resU = np.zeros((10,1,u.shape[2]))

for i in range(10):
    pert = h[i]*np.random.rand(*u.shape)
    fPerturbedu = interpScalar.eval(x,u+pert)
    resU[i,0,:] = np.linalg.norm(fPerturbedu - (f + mult(dfdu,pert)), axis = 0)
    
maxResU = np.max(resU, axis = 2); medResU = np.median(resU, axis = 2);

plt.figure()
plt.loglog(h, medResU,'-o', label =  "median of the residual")
plt.loglog(h, maxResU,'-o', label =  "maximum of the residual")
plt.loglog(h, h**2,'k--', label =  "expected decay $(h^2)$")
plt.legend(loc = "best"); plt.grid()
plt.xlabel("$h$"); plt.ylabel("Euclidian norm of Taylor remainder")
plt.title("Taylor remainder with respect to $u$")
plt.show()

#%% Check Taylor expansion - x

resX = np.zeros((10,1,u.shape[2]))
l = list(x.keys())[0]
xPert = x.copy()

for i in range(10):
    pert = h[i]*np.random.rand(*x[l].shape)
    xPert[l] = x[l] + pert
    fPerturbedx = interpScalar.eval(xPert,u)
    pert = np.reshape(pert.T, (2,1,-1))
    resX[i,0,:] = np.linalg.norm(fPerturbedx - (f + mult(dfdx[l],pert)), axis = 0)

maxResX = np.max(resX, axis = 2); medResX = np.median(resX, axis = 2);

plt.figure()
plt.loglog(h, medResX,'-o', label =  "median of the residual")
plt.loglog(h, maxResX,'-o', label =  "maximum of the residual")
plt.loglog(h, h**2,'k--', label =  "expected decay $(h^2)$")
plt.legend(loc = "best"); plt.grid()
plt.xlabel("$h$"); plt.ylabel("Euclidian norm of Taylor remainder")
plt.title("Taylor remainder with respect to $x$")
plt.show()

#%% 2) Vector case
#%% a) Some operations

from hiped.utils import mult

dimInput = 2
dimOutput = 2

f1v = hp.VertexFunction(label = "f1v", f = lambda u : u**2, dfdu = lambda u : 2*u * np.eye(2)[:,:,None], dimInput= dimInput, dimOutput = dimOutput)
f2v = hp.VertexFunction("f2v", lambda u : mult(np.array([[1,2],[3,4]]), u), lambda u : np.array([[1,2],[3,4]])[:,:,None] * np.ones(u.shape), dimInput, dimOutput)
f3v = hp.VertexFunction("f3v", lambda u : 0.1*u, lambda u : 0.1 * np.eye(2)[:,:,None] * np.ones(u.shape), dimInput, dimOutput)

fMat =  hp.VertexFunction("fMat", lambda u : np.array([[1,2],[3,4]]).reshape(2,2,1)*np.ones(u.shape), lambda u : np.zeros(u.shape) * np.zeros((2,2,1)), dimInput, (2,2))
fConst =  hp.VertexFunction("fConst", lambda u : np.ones((1,1,1))*np.ones(u.shape), lambda u : np.zeros((1,1,1))*np.zeros(u.shape), dimInput, 1)
fConst2 =  hp.VertexFunction("fConst2", lambda u : np.ones((2,1,1))*np.ones(u.shape), lambda u : np.zeros((2,2,1))*np.zeros(u.shape), dimInput, 2)

f4v = fMat @ f1v * (f3v+f2v) / 2 # only multiplication by constant matrix is supported
print(f4v)
f5v = 3* f1v.innerProduct(f2v+f3v)**f1v 
print(f5v)

#%% b) Interpolation

n = 100
penal = hp.Penalization("simp", 2)
interpVector= hp.Interpolation(domain = hp.Domain(5), children = [f1v,f2v,f3v,f4v,f5v],
                               label = "interpVector", penalization = penal)


x = interpVector.setInitialVariable(n, "rand", radius = 1) # initialization of the variables
x = interpVector.projection(x) # projection onto the domain
u = np.random.rand(dimInput,1,n)
plt.figure()
interpVector.plot(x)
plt.show()

#%% c) Evaluation

w, dwdx = interpVector.evalBasisFunction(x)
f = interpVector.eval(x, u, w) # value
dfdu =  interpVector.evaldu(x, u, w) # derivative w.r.t a
dfdx =  interpVector.evaldx(x, u, w, dwdx) # derivative w.r.t x

#%% Check Taylor expansion - u

h = np.logspace(-8,-2,10) # test for 10 different h
resU = np.zeros((10,1,u.shape[2]))

for i in range(10):
    pert = h[i]*np.random.rand(*u.shape)
    fPerturbedu = interpVector.eval(x,u+pert)
    resU[i,0,:] = np.linalg.norm(fPerturbedu - (f + mult(dfdu,pert)), axis = 0)
    
maxResU = np.max(resU, axis = 2); medResU = np.median(resU, axis = 2);

plt.figure()
plt.loglog(h, medResU,'-o', label =  "median of the residual")
plt.loglog(h, maxResU,'-o', label =  "maximum of the residual")
plt.loglog(h, h**2,'k--', label =  "expected decay $(h^2)$")
plt.legend(loc = "best"); plt.grid()
plt.xlabel("$h$"); plt.ylabel("Euclidian norm of Taylor remainder")
plt.title("Taylor remainder with respect to $u$")
plt.show()

#%% Check Taylor expansion - x

resX = np.zeros((10,1,u.shape[2]))
l = list(x.keys())[0]
xPert = x.copy()

for i in range(10):
    pert = h[i]*np.random.rand(*x[l].shape)
    xPert[l] = x[l] + pert
    fPerturbedx = interpVector.eval(xPert,u)
    pert = np.reshape(pert.T, (2,1,-1))
    resX[i,0,:] = np.linalg.norm(fPerturbedx - (f + mult(dfdx[l],pert)), axis = 0)

maxResX = np.max(resX, axis = 2); medResX = np.median(resX, axis = 2);

plt.figure()
plt.loglog(h, medResX,'-o', label =  "median of the residual")
plt.loglog(h, maxResX,'-o', label =  "maximum of the residual")
plt.loglog(h, h**2,'k--', label =  "expected decay $(h^2)$")
plt.legend(loc = "best"); plt.grid()
plt.xlabel("$h$"); plt.ylabel("Euclidian norm of Taylor remainder")
plt.title("Taylor remainder with respect to $x$")
plt.show()

